### DISCUSSION TITLE
SQL Query Examples

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#136] [Q&A] [answered]
I want to make an SQL query with the following goal
select the documents to return (list provided) and if they are borrowed put the name of the borrower and the expected return date
little more if the documents are not borrowed put "available" in the borrower's name column

Je veux faire une requête SQL avec le but suivant
sélectionner les documents à rendre (liste fournie) et s'ils sont empruntés mettre le nom de l'emprunteur et la date de retour prévue
un petit plus si les documents ne sont pas empruntés mettre "disponible" dans la colonne du nom de l'emprunteur

SELECT callnumber, title, author, isbn13,  last_name, duedate  
FROM book, member, item_borrower 
WHERE accession_number IN ('7098','7095','7090')
INNER JOIN  book ON item_borrower.item_oid = book.myoid
INNER JOIN member ON item_borrower.memberid = member.memberid

I can't get any results ?

**Comment by meteos77** [comment]
to select several documents in the database with specific fields:
pour sélectionner plusieurs documents dans la base avec les champs spécifiques :

custom query:
SELECT callnumber, title, author, isbn13 FROM book WHERE accession_number IN ('7056','7057','7058')

the result is ok but in the table the first field (callnumber) is always invisible (uncheck) in the presentation is it voluntary ?

I created an issue by mistake, I wanted this answer to become a new topic in the SQL query examples I don't know how to do it?

**Comment by textbrowser** [comment]
Issue is fine. :)

**Reply by meteos77** [reply]
I am trying to set up BiblioteQ with the different actions that have to be performed at the library

**Reply by textbrowser** [reply]
I meant submitting a ticket is good.

**Reply by meteos77** [reply]
I'll need a little help on the SQL queries because the first one returns an error with the INNER
if I delete the useless INNER JOINT ?

it returns me the right result but in 12 copies
[2-2021-11-27_liste-47_achat_nouveautés.sqlite.zip](https://github.com/textbrowser/biblioteq/files/7693346/2-2021-11-27_liste-47_achat_nouveautes.sqlite.zip)
?

**Comment by textbrowser** [comment]
> SELECT callnumber, title, author, isbn13, last_name, duedate
> FROM book, member, item_borrower
> WHERE accession_number IN ('7098','7095','7090')
> INNER JOIN book ON item_borrower.item_oid = book.myoid
> INNER JOIN member ON item_borrower.memberid = member.memberid

This query?

**Reply by meteos77** [reply]
select the documents to return (list provided) and if they are borrowed put the name of the borrower and the expected return date
little more if the documents are not borrowed put "available" in the borrower's name column

**Comment by textbrowser** [comment]
Great.

**Comment by textbrowser** [comment]
How about "All Reserved" category?

**Comment by textbrowser** [comment]
Due Date and Reservation Date are included in All Reserved.

**Reply by meteos77** [reply]
this does not answer my request

I have a list of documents to take out of the library
before making the physical removal of the list, I would like to modify this list by indicating; this one is in the shelves thus to seek
this one is borrowed so no need to go and see in the library (this list is composed of 300 or 400 documents)

**Reply by textbrowser** [reply]
Right. All Reserved indicates the items that are loaned out. BQ will not allow you to remove a reserved item.

**Reply by textbrowser** [reply]
All Reserved indicates that the items are on loan. You can review each item's physical correctness by looking at the All Reserved list and comparing it to the physical shelves. If the item is on the shelf, select it in BQ and specify that it has been returned. If the item is actually missing from the shelf, repeat the process of returning it and then remove it from the list.

**Comment by textbrowser** [comment]
You'd like All Reserved for specific items, yes?

**Comment by textbrowser** [comment]
For the database you shared, I see 15 items in All Reserved.

**Comment by textbrowser** [comment]
I will look at All Reserved and adjust it to retrieve only those items.

**Comment by textbrowser** [comment]
Then you can curtail the columns.

**Comment by meteos77** [comment]
I have 3 books to get out, 1 of which has been released by someone
in the list of the 3 to go out, the list is of 3 books 1 out with the name of the borrower and the 2 others are available so to search in the library

**Reply by textbrowser** [reply]
So I am looking at your database in All Reserved. Which items do you want to remove?

**Reply by textbrowser** [reply]
BQ will prevent you from removing reserved items. You simply have to simulate a return process and then remove them.

**Reply by meteos77** [reply]
I don't want to delete anything at this stage

let's take another list
SELECT callnumber, title, author, isbn13, last_name, duedate FROM book, member, item_borrower WHERE accession_number IN ('7057', '7058', '7060')

it should tell me the 7057 is borrowed by Name1
7058 available
7060 available

**Reply by textbrowser** [reply]
This is a long list. I'm not understanding why All Reserved is not sufficient? It includes the information you're searching for except that it lists all reserved items.

**Reply by textbrowser** [reply]
OK, I'll look at using your query. :P

**Reply by textbrowser** [reply]
Which requires my time. :P All Reserved requires your time and it presents a superset of the information you're in need of. Fine, I'll help because we're at a misunderstanding.

**Reply by meteos77** [reply]
but the unborrowed items are not present in the list even though I need to get them out too

**Reply by textbrowser** [reply]
All Reserved only includes reserved items.

**Comment by textbrowser** [comment]
After lunch.

**Reply by meteos77** [reply]
ok thanks, I'm not used to SQL except for a simple select
with this complex example it will give a good basis for me to learn how to do others :-)

**Comment by textbrowser** [comment]
Did you read the note on top of Custom Query? :P

**Comment by textbrowser** [comment]
"... intimate knowledge of SQL ..."

**Comment by textbrowser** [comment]
Expectations, expectations. Custom Query should be removed! :)

**Reply by meteos77** [reply]
you're the one who sold me on SQL so I'm trying to take advantage of it :-)

**Reply by textbrowser** [reply]
Sure. However, it is you who must invest in yourself to learn it.

**Comment by textbrowser** [comment]
Query:

select book.callnumber, book.title, book.author, book.isbn13, member.last_name, item_borrower.duedate from member, book left join item_borrower on book.myoid = item_borrower.item_oid where book.accession_number in  ('7098','7095','7090') and item_borrower.memberid = member.memberid

produces:

R NOT|PREMIER SANG - PRIX RENAUDOT 2021|NOTHOMB AMELIE|9782226465382|Nom 2|12/02/2021

**Reply by meteos77** [reply]
this is not the expected result
I want all the documents in the list but I want to get from the database the documents that are out with their borrowers

the 3 books must appear

**Reply by textbrowser** [reply]
But only one of those books is on loan.

**Reply by textbrowser** [reply]
The 7098 is reserved. 7090 and 7095 are not reserved.

**Reply by textbrowser** [reply]
You want 7090 and 7095 to appear on the list?

**Reply by textbrowser** [reply]
I don't know if SQLite supports RIGHT joins.

**Reply by textbrowser** [reply]
I need to create a UNION.

**Comment by textbrowser** [comment]
select book.callnumber, book.title, book.author, book.isbn13, member.last_name, item_borrower.duedate from member, book left join item_borrower on book.myoid = item_borrower.item_oid where item_borrower.memberid = member.memberid;

R MCD|Betty|Tiffany McDaniel|9782351782453|Nom 2|11/30/2021
R HAW|Celle qui brûle|Paula Hawkins|9782355848858|Nom 2|11/30/2021
N/A|DOUCE, DOUCE VENGEANCE|JONASSON JONAS|9782258193475|Nom 2|11/30/2021
N/A|ENVOYE UN PEU SPECIAL|BLANC-GRAS JULIEN|9782234089402|Nom 2|11/30/2021
R FOM|ALMA - LE VENT SE LEVE|FOMBELLE/PLACE|9782075139106|Nom 1|11/30/2021
N/A|JE SUIS L'ABYSSE|CARRISI DONATO|9782702182895|Nom 2|11/30/2021
R COL|LA CHARMANTE LIBRAIRIE DES FLOTS TRANQUILLES|COLGAN JENNY|9782810431717|Nom 1|11/30/2021
N/A|L'ANGE DECHU|Chris Brookmyre|9791022611169|Nom 2|12/01/2021
N/A|L'HOMME DES HAUTES COMBES|BOUCHET JEAN-PAUL|9782953047363|Nom 2|12/01/2021
N/A|LA CLAUSE PATERNELLE|KHEMIRI JONAS HASSEN|9782330150990|Nom 1|12/01/2021
N/A|LA JEUNE FEMME ET LA MER|MEURISSE CATHERINE|9782205089691|Nom 1|12/01/2021
N/A|MOHICAN|FOTTORINO ERIC|9782072941740|Nom 2|12/02/2021
N/A|MON MARI|VENTURA MAUD|9782378802417|Nom 2|12/02/2021
N/A|NE M'OUBLIE PAS|GARIN ALIX|9782803676231|Nom 2|12/02/2021
R NOT|PREMIER SANG - PRIX RENAUDOT 2021|NOTHOMB AMELIE|9782226465382|Nom 2|12/02/2021

**Reply by textbrowser** [reply]
All Reserved produces the same number of results.

**Reply by meteos77** [reply]
OK it is the equivalent of All Borrowings but not equal to the demand

**Reply by meteos77** [reply]
expected result

![attendu](https://user-images.githubusercontent.com/56431039/145613575-dffe7836-c300-4b7e-bb34-d161998a8e14.png)

**Reply by textbrowser** [reply]
Right. SQLite does not support RIGHT joins so it has to be emulated. Here is a proper request:

"I would like to see some columns from book, some columns from item_borrower, and some columns from member for a specific set of books. If the items are not borrowed, I'd like to see those items regardless."

**Reply by textbrowser** [reply]
The sample query does not contain such information. Just say I want to see the books even if they're not borrowed.

**Reply by textbrowser** [reply]
:P Which is another query altogether.

**Reply by textbrowser** [reply]
In PostgreSQL, this would be easy.

**Reply by textbrowser** [reply]
I reread your initial request. Now it's more clear. :)

**Reply by meteos77** [reply]
Of course I will dig but some examples made by a specialist are always very instructive
in some minutes I have notions SQL RIGHT Joins; UNION to deepen, if I read the documentation, to make the connection of what I wish to make with the base biblioteQ will become impossible.

**Reply by textbrowser** [reply]
I have a document with custom queries for my own curiosities.

**Reply by textbrowser** [reply]
Indeed, the basic categories fulfill a very small set of queries. Hence the custom queries for deeper meaning into the data.

**Reply by meteos77** [reply]
why not share it?

**Reply by textbrowser** [reply]
PostgreSQL and they're not meaningful for anyone else.

**Comment by textbrowser** [comment]
OK, let me gather the proper query and move on to something else. :)

**Comment by textbrowser** [comment]
I will also improve the opening of Item View ... Details for custom queries. That's also important.

**Comment by textbrowser** [comment]
SELECT book.callnumber, book.title, book.author, book.isbn13, member.last_name, item_borrower.duedate
FROM book LEFT JOIN item_borrower ON book.myoid = item_borrower.item_oid, member 
WHERE book.accession_number IN  ('7098','7095','7090') AND item_borrower.memberid = member.memberid
UNION
SELECT book.callnumber, book.title, book.author, book.isbn13, 'Disponible' AS last_name, item_borrower.duedate
FROM book LEFT JOIN item_borrower ON book.myoid = item_borrower.item_oid
WHERE book.accession_number IN ('7098','7095','7090') and item_borrower.item_oid IS NULL

R DES|MON MAITRE ET MON VAINQUEUR|DESERABLE F-H.|9782072900945|Disponible|
R GRI|LES POSSIBLES|GRIMALDI VIRGINIE|9782363606600|Disponible|
R NOT|PREMIER SANG - PRIX RENAUDOT 2021|NOTHOMB AMELIE|9782226465382|Nom 2|12/02/2021

**Reply by textbrowser** [reply]
Happy?

**Reply by meteos77** [reply]
thank you very much, I will be able to study how you did it

**Reply by meteos77** [reply]
you are Santa Claus :-)

**Reply by textbrowser** [reply]
Now let me see if I can adjust the opening of those items in BQ. Custom queries may not contain the required columns.

**Reply by meteos77** [reply]
I didn't understand everything but your query does the job well and BiblioteQ scores 1 point in the simplicity of managing the library

**Reply by textbrowser** [reply]
Double-click on one of the items. BQ will not be able to open it.

**Reply by meteos77** [reply]
ok ok
can we put a variable in the query containing the list of accession_number to put it only once?

**Reply by textbrowser** [reply]
With a temporary table, yes.

**Reply by meteos77** [reply]
ok I'm looking for

**Reply by textbrowser** [reply]
It's a lot of work and may not work. See PRAGMA TEMP_STORE.

**Comment by textbrowser** [comment, marked as answer]
SQL is a rich language. It's like LISP. I spend more time thinking about them than typing.

**Comment by textbrowser** [comment]
SELECT book.callnumber, book.title, book.author, book.isbn13, book.myoid, book.type, member.last_name, item_borrower.duedate
FROM book LEFT JOIN item_borrower ON book.myoid = item_borrower.item_oid, member
WHERE book.accession_number IN ('7098','7095','7090') AND item_borrower.memberid = member.memberid
UNION
SELECT book.callnumber, book.title, book.author, book.isbn13, book.myoid, book.type, 'Disponible' AS last_name, item_borrower.duedate
FROM book LEFT JOIN item_borrower ON book.myoid = item_borrower.item_oid
WHERE book.accession_number IN ('7098','7095','7090') and item_borrower.item_oid IS NULL

The new Custom Query dialog will include a reminder for myoid and type columns.

**Reply by textbrowser** [reply]
Committed.
